lib/improver/convection.py

```diff
@@ -0,0 +1,232 @@
+                  'radii: {}; fuzzy_factor {}; '
+                  'below_thresh_ok: {}; lead_times: {}; '
+                  'unweighted_mode: {}; ens_factor: {}; '
+                  'use_adjacent_grid_square_differences: {}>')
+        return result.format(
+            self.lower_threshold, self.higher_threshold,
+            self.neighbourhood_method, self.radii, self.fuzzy_factor,
+            self.below_thresh_ok, self.lead_times, self.unweighted_mode,
+            self.ens_factor, self.use_adjacent_grid_square_differences)
+
+    def _calculate_convective_ratio(self, cube):
+        """
+        Calculate the convective ratio by:
+        1. Threshold the input cube using the lower_threshold and
+           apply neighbourhood processing.
+        2. Threshold the input cube using the higher_threshold and
+           apply neighbourhood processing.
+        3. Calculate the convective ratio by:
+           higher_threshold_cube / lower_threshold_cube.
+           For example, the higher_threshold might be 5 mm/hr, whilst the
+           lower_threshold might be 0 mm/hr.
+
+        The Convective Ratio can have the following values:
+            * A non-zero fractional value, indicating that both the higher
+              and lower thresholds were exceeded.
+            * A zero value, if the lower threshold was exceeded, whilst the
+              higher threshold was not exceeded.
+            * A NaN value (np.nan), if neither the higher or lower thresholds
+              were exceeded, such that the convective ratio was 0/0.
+
+        Args:
+            cube : Iris.cube.Cube
+                The cube from which the convective ratio will be calculated.
+
+        Returns:
+            convective_ratio : Iris.cube.Cube
+                Cube containing the convective ratio
+
+        Raises:
+            ValueError: If a value of infinity or a value greater than 1.0
+                        are found within the convective ratio.
+
+        """
+        threshold_dict = {}
+
+        for threshold in [self.lower_threshold, self.higher_threshold]:
+            threshold_cube = (
+                BasicThreshold(
+                    threshold, fuzzy_factor=self.fuzzy_factor,
+                    below_thresh_ok=self.below_thresh_ok).process(cube.copy()))
+            neighbourhooded_cube = NeighbourhoodProcessing(
+                self.neighbourhood_method, self.radii,
+                lead_times=self.lead_times,
+                unweighted_mode=self.unweighted_mode,
+                ens_factor=self.ens_factor).process(threshold_cube)
+            threshold_dict[threshold] = neighbourhooded_cube
+
+        # Ignore runtime warnings from divide by 0 errors.
+        with np.errstate(invalid='ignore', divide='ignore'):
+            convective_ratio = (
+                threshold_dict[self.higher_threshold] /
+                threshold_dict[self.lower_threshold])
+
+        infinity_condition = np.sum(np.isinf(convective_ratio.data)) > 0.0
```

Comment:
I think we need to have a chat about what the values should be in certain situations,like where there is no precipitation and where the first threshold is not exceeded and agree on it. I'm not sure this is quite right at the moment, but we probably need to discuss it, because I'm not sure what the right things are.